### DISCUSSION TITLE
Beam elements with open sections

### DISCUSSION THREAD

**Discussion by xyont** [TLCFEM/suanPan#156] [Ideas]
Hi,

Probably this is fairly new, common latest improvement of beam element are for open section. It's advanced in stability analysis purpose. Hopefully these element types are interested to implement and will be available in the future of SuanPan.

Best,

**Comment by TLCFEM** [comment]
I am always interested in this topic, such as the generalised beam theory that shows great capability in terms of modelling local warping and torsion. But I have never found any detail on implementation. Do you have any reference documents available, or have any elements in mind?

**Reply by xyont** [reply]
Thank you for interested in, actually i'm only reading the capabilities of beam element with open sections. It has given promise in direct analysis and design. 

Some reported which almost complete in documentation are from [Rinchen](https://ses.library.usyd.edu.au/handle/2123/18882) of University of Sydney (2018) and [Chen Liang](https://theses.lib.polyu.edu.hk/handle/200/12365) of Hong Kong Polytechnic University (2022)

**Reply by TLCFEM** [reply]
The provided two references present elements that are not very complex. They are effectively conventional beam elements with additional torsion DoF. This shall be relatively easy to implement, in the context of elastic analysis only. For stability analysis, within the element, it is first order, buckling analysis can be achived by mesh refinement with corotational transformation.

Elements of this type may not be applicable to plastic analyses, as the consitutitve laws for torsion/shear are unclear. But implementing an elastic element should not be very difficult to do.

**Comment by xyont** [comment]
Also, thanks for remembering inelastic analysis limitation of beam open section when subjected to torsion/shear. The documentation also shown, inelastic analysis is calculated for axial forces, major and minor bending only. It's based on section yield surface (P, Mx, My) and load ratio to reduced element stiffness.

**Comment by TLCFEM** [comment]
I have added an elastic 3D beam element `EB31OS` with both linear and corotational transformations in #158. I have done some simple tests and the convergence appears to be correct. If you can help testing it, it will be great.

Check the syntax [here](https://tlcfem.github.io/suanPan-manual/3.2/Library/Element/Beam/EB31OS/), and the example files `/Example/Element/EB31OS.supan` and `/Example/Element/EB31OSNL.supan`.

I do not see a strong need to implement plastic models at material point level with strains computed from section deformation given that:
1. When torsion, warping, instability are of concern, the FLT buckling is mainly elastic.
2. Unless using 2D material models (which would still introduce unrealistic assumptions no matter how), separate considerations of plasticity developments of normal and shear stresses do not sound correct to me.

Liang's thesis uses yield surface on section response, which is simpler than the strain/stress level approach, but I am not a fan of that stiffness reduction method. In this regard, I do see the potential of introducing elastic torsion and warping into the model described in [this](https://github.com/TLCFEM/nm-formulation) paper.

**Reply by TLCFEM** [reply]
> In this regard, I do see the potential of introducing elastic torsion and warping into the model described in [this](https://github.com/TLCFEM/nm-formulation) paper.

I think this would be a better approach as it is possible to achieve quadratic convergence.

**Reply by xyont** [reply]
Great to hear when it is possible to include plasticity of open section beam element, even still have some limitation in torsion and warping treated being elastic.

**Comment by xyont** [comment]
Indeed, it seems OpenSees can capture better in behavior than yield surface approach as in Mastan2 since torsional being included in plasticity analysis.

Link of OpenSees implementation codes can be found [here](https://github.com/xinlong-du/OpenSees-3.0.0), and both paper references available at NSF.

**Comment by TLCFEM** [comment]
I've added a general beam element `B31OS`, and the corresponding `Cell3DOS` section to be used to build arbitrary sections. Some preliminary testing appears to be fine to me, see examples [here](https://tlcfem.github.io/suanPan-manual/3.2/Example/Structural/Statics/thin-walled-section/).

I revised the torsion strain formulation and split the tangent shear strain into two $\gamma_{12}$ and $\gamma_{13}$. And a universal wrapper `OS146` is added to wrap whatever 3D material model into proper one to be used. By such, I believe it can be used for whatever sections, thin-/thick-walled open/close sections.

Let me know if you are willing/available to further test it. If so I can give you more info, otherwise I'll leave it as it is for the moment.

**Comment by xyont** [comment]
unfortunately, i'm not use and familiar with the Python library used. Is the mesh generated by the library need to be used in SuanPan? 

In case of arbitrary section shapes, i'm also using another FE software but tabulated later based on variable dimensions and categories. So, the section properties can be read from table when it needed.

Perhaps generic input file format required is documented, not only to depend on specific external software such as Python library in examples.

**Reply by TLCFEM** [reply]
Basically warping function and its derivatives need to be provided. Doesn't really matter how those are computed. The python library is accessible to everyone. I presume other tools may also allow to export warping function at nodes in one way or another.
